File: code/java/1353.java
// https://leetcode.com/problems/maximum-number-of-events-that-can-be-attended/
class Solution {
    int search(int[] nextDay, int day) {
        // Union-Find search function
        while (nextDay[day] != day) {
            nextDay[day] = nextDay[nextDay[day]];
            day = nextDay[day];
        }
        return day;
    }

    public int maxEvents(int[][] events) {
        Arrays.sort(events, (a, b) -> a[1] - b[1]);
        // Initialize disjoint set
        int[] nextDay = new int[events[events.length - 1][1] + 2];
        for (int d = 0; d < nextDay.length; d++)
            nextDay[d] = d;

        int res = 0;
        for (int[] event : events) {
            int day = search(nextDay, event[0]);
            if (day <= event[1]) {
                res++;
                nextDay[day] = search(nextDay, day + 1);
            }
        }
        return res;
    }
}

File: code/python/1353.py
# https://leetcode.com/problems/maximum-number-of-events-that-can-be-attended/
class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        def find(day): // Union-Find search function
            while next_day[day] != day:
                next_day[day] = next_day[next_day[day]]
                day = next_day[day]
            return day

        events.sort(key=lambda x: x[1])
        next_day = [x for x in range(events[len(events) - 1][1] + 2)] // Initialize disjoint set
        res = 0
        for start, end in events:
            day = find(start)
            if day <= end:
                res += 1
                next_day[day] = find(day + 1)
        return res
